Topics/Tech_Stacks/cuda.md

```diff
@@ -0,0 +1,260 @@
+# CUDA (Compute Unified Device Architecture)
+
+## Table of Contents:
+#### [Introduction]()
+#### [Installation]()
+#### [A brief primer to GPU architecture]()
+#### [Writing our first kernel]()
+#### [Optimization techniques]()
+
+## Introduction:
+CUDA, short for Compute Unified Device Architecture, is a proprietary toolkit offered for NVIDIA GPUs that takes advantage of the massive amount of parallelism GPUs offer. In today's landscape, algorithmic complexity and computational theory will only take us so far, and bringing down an algorithm from $O(n^2)$ to $O(n)$ runtime complexity for example will surely help us improve performance, but what if we could assign $n$ units to split up work on our algorithm at a massive scale? Additionally, what if we run into computational problems where there is no good solution? Like the traveling salesman problem? Or gradient descent in machine learning? Or matrix multiplication? Despite our best efforts in reducing these problems computational costs either theoretically or practically through optimization techniques at the CPU level, they are still too costly to perform at a large scale. This is where GPUs come into play. Instead of having one horse pulling a cart, what if you had a thousand chickens? Or ten thousand chickens? Maybe even a million chickens? GPUs are like the million chickens pulling a cart. By having poorer and less versatile performing cores at an individual level than CPUs, we are able to fit thousands, even tens of thousands of cores on a GPU which can compute parallel tasks at a massive scale.
```

Comment:
When listing examples that are hard to reduce the runtime complexity and suggesting that parallel programming at the GPU level could be an alternative way to reduce runtime cost, it could be better to also state that the computation involved are rather simple. The instruction sets designed into CPUs are likely much more than what it needs to perform the required computations.